feat: add shuttle extensions cli commands

kjuulh opened the pull request:
This adds a basic command palette for shuttle such as:

- `shuttle ext update`
- `shuttle ext init`
- `shuttle ext install`

This also right now uses a skeleton implementation of a central ExtensionManager which is supposed to orchestrate these extensions. The call to the extensions will be done in a later PR

Builds on: #219

kjuulh committed:
feat: adds basic commands

Signed-off-by: Kasper J. Hermansen <contact@kjuulh.io>

kjuulh committed:
feat: add some more description

Signed-off-by: Kasper J. Hermansen <contact@kjuulh.io>

kjuulh commented:
@lunarway/squad-aura I really like to get this merged before hackathon. 

I am opening up the PRs now (stacked), They all live pretty much in the same package, so it'd be easy to squash them down before merging to main, if we want to move this to a separate tool in the future.

mahlunar commented on cmd/ext.go:
```diff
@@ -0,0 +1,59 @@
+)
+
+func newExtCmd() *cobra.Command {
+	extManager := extensions.NewExtensionsManager("some registry")
```
It seems wrong to hard code a string here?

kjuulh replied:
Will be replace in a further pr. Won't fix

mahlunar commented on cmd/ext.go:
```diff
@@ -0,0 +1,59 @@
+func newExtInstallCmd(extManager *extensions.ExtensionsManager) *cobra.Command {
+	cmd := &cobra.Command{
+		Use:  "install",
+		Long: "Install ensures that extensions already known about are downloaded and available",
```
Something is off in this text? "known about" could be "registered"?

kjuulh replied:
I'll skip the entire known about and reigsterered, as we haven't implemented that part yet. (i.e. you can't choose currently) You're only getting the registry.

kjuulh committed:
docs: reword

Signed-off-by: Kasper J. Hermansen <contact@kjuulh.io>

Commit:
feat: add git extension registry for shuttle extensions (#221)

* feat: add registry and global storage

Signed-off-by: Kasper J. Hermansen <contact@kjuulh.io>

* feat: remove bullet

Signed-off-by: Kasper J. Hermansen <contact@kjuulh.io>

* feat: remember to initiate install

Signed-off-by: Kasper J. Hermansen <contact@kjuulh.io>

* feat: make sure to check permissions

Signed-off-by: Kasper J. Hermansen <contact@kjuulh.io>

* chore: make registry private

Signed-off-by: Kasper J. Hermansen <contact@kjuulh.io>

* feat(extensions): with git operations (#222)

* feat(extensions): with git operations

Quite simply implements fetch and clone to make sure we can pull a registry

Signed-off-by: Kasper J. Hermansen <contact@kjuulh.io>

* feat: add git registry

Signed-off-by: Kasper J. Hermansen <contact@kjuulh.io>

* feat(extensions): add install extensions from registry (#223)

* "feat(extensions): add install extensions from registry

This pr includes everything up to actually downloading the extension from the upstream registry

Such as pulling an example registry entry

```json
{
  "name": "SuperExtender",
  "description": "Enhances your experience with superpowers.",
  "version": "1.0.3",
  "downloadUrls": [
    {
      "architecture": "amd64",
      "os": "linux",
      "url": "https://example.com/linux/amd64/superextender.tar.gz",
      "checksum": "efgh5678"
    },
    {
      "architecture": "x86",
      "os": "windows",
      "url": "https://example.com/windows/x86/superextender.zip",
      "checksum": "ijkl91011"
    },
    {
      "architecture": "arm64",
      "os": "darwin",
      "url": "https://example.com/windows/x86/superextender.zip",
      "checksum": "ijkl91011"
    }
  ]
}
```

Which will match on the current os using $GOOS and $GOARCH set by golang itself. We may however have to depend on some other variables if go is not installed by I believe it should work anyway"

* feat: print error

Signed-off-by: Kasper J. Hermansen <contact@kjuulh.io>

* fix: review comments, wording etc.

Signed-off-by: Kasper J. Hermansen <contact@kjuulh.io>

* feat(extensions): add github release downloader (#224)

* feat(extensions): add github release downloader

This specifically adds a downloader for github releases it uses either the

SHUTTLE_EXTENSIONS_GITHUB_ACCESS_TOKEN or GITHUB_ACCESS_TOKEN for access to the releases

The downloader expects a full url from the registry as seen in a previous pr

Signed-off-by: Kasper J. Hermansen <contact@kjuulh.io>

* chore: reword bearer

Signed-off-by: Kasper J. Hermansen <contact@kjuulh.io>

* feat(extensions): add execute command (#225)

* feat(extensions): add execute command

this adds Extensions to the global cmd.go resulting in another section showing up when doing a shuttle --help

To execute an extension simply shuttle myExtension where myExtension is the name of the downloaded extension. All args are passed to the child as well

Signed-off-by: Kasper J. Hermansen <contact@kjuulh.io>

* feat: remember to add error

Signed-off-by: Kasper J. Hermansen <contact@kjuulh.io>

* feat: without empty line

Signed-off-by: Kasper J. Hermansen <contact@kjuulh.io>

* feat: adds github remote registry index (#226)

* feat: with github registry

Signed-off-by: Kasper J. Hermansen <contact@kjuulh.io>

* fix: shuttle extensions

Signed-off-by: Kasper J. Hermansen <contact@kjuulh.io>

* feat: can download private files

Signed-off-by: Kasper J. Hermansen <contact@kjuulh.io>

* feat: remove fluff

Signed-off-by: Kasper J. Hermansen <contact@kjuulh.io>

* feat: fix review comments

Signed-off-by: Kasper J. Hermansen <contact@kjuulh.io>

* feat: it needs to implement the functions

Signed-off-by: Kasper J. Hermansen <contact@kjuulh.io>

---------

Signed-off-by: Kasper J. Hermansen <contact@kjuulh.io>

---------

Signed-off-by: Kasper J. Hermansen <contact@kjuulh.io>

---------

Signed-off-by: Kasper J. Hermansen <contact@kjuulh.io>

---------

Signed-off-by: Kasper J. Hermansen <contact@kjuulh.io>

---------

Signed-off-by: Kasper J. Hermansen <contact@kjuulh.io>

---------

Signed-off-by: Kasper J. Hermansen <contact@kjuulh.io>
